Fix Bug in MixedPrecisionLossScaleOptimizer

zhuzilin opened the pull request:
This is a PR from JIZHI, the AI platform in Tencent.

This pr mainly fixed the error that when disable eager execution and use the `MixedPrecisionLossScaleOptimizer` as the opt for keras model will trigger
```bash
FailedPreconditionError: Error while reading resource variable current_loss_scale from Container: localhost. This could mean that the variable was uninitialized. Not found: Resource localhost/current_loss_scale/N10tensorflow3VarE does not exist.
	 [[{{node training/ReadVariableOp_1}}]]
```
The error is able to be replicated in tf-nightly, see gist [here](https://colab.research.google.com/gist/zhuzilin/f116e9689e2611333719fa753af31599/amp-opt-bug.ipynb)

The reason for this error is that the `MixedPrecisionLossScaleOptimizer` failed to pass the variables defined in its base optimizer and loss scale to keras backend and initialize them.

This PR fixed it by overwriting the `variables()` function, which is the function called to initialize variables.

Thank you for your time on reviewing this PR.

zhuzilin committed:
Fix bug in MixedPrecisionLossScaleOptimizer

reedwm commented on tensorflow/python/training/experimental/loss_scale.py:
```diff
@@ -197,6 +197,10 @@ def from_config(cls, config):
     """Creates the LossScale from its config."""
     return cls(**config)
 
+  def variables(self):
```
I would remove this change, and just access `loss_scale._weights` in loss_scale_optimizer.py. Use the `# pylint: disable=protected-access` comment to disable the lint error. The reason is that I am unsure if we want to expose this function as a public API.

zhuzilin replied:
Removed

reedwm reviewed (changes requested):
Can you add a test? Simply ensuring you can call `model.fit()` on a Keras model with a MixedPrecisionLossScaleOptimizer is sufficient.

zhuzilin committed:
remove the added api on loss_scale

zhuzilin commented:
@reedwm Thank you for your review! I've removed the api exposed on loss_scale but not sure where should I put the test? Should it be in optimizers_test.py in keras or the loss_sacle_optimizer_test.py next to the file I modified?

reedwm commented:
Add the test to optimizers_test.py, since we are trying to remove any dependencies from TF to Keras. optimizers_test.py already has a test that runs Model.fit with a non-Keras AdamOptimizer, so you can add a new test that runs with a MixedPrecisionLossScaleOptimizer.

zhuzilin committed:
add test for MixedPrecisionLossScaleOptimizer

zhuzilin commented:
@reedwm I've added the test. Could you have a look? Thank you!
By the way, the long import line can pass the pylint since there is no resrict on the length of import.

reedwm commented on tensorflow/python/training/experimental/loss_scale_optimizer.py:
```diff
@@ -243,3 +243,8 @@ def _resource_apply_sparse(self, grad, handle, indices):
+
+  def variables(self):
+    """Returns the variables of the Optimizer."""
+    return self._optimizer.variables() + \
```
Instead of using a backslash here, wrap the entire expression in parenthesis:

```
return (self._optimizer.variables() +
        list(self._loss_scale._weights.values())) # pylint: disable=protected-access
```

[Bashslashes aren't allowed by the style guide](http://google.github.io/styleguide/pyguide.html#32-line-length).

zhuzilin replied:
changed

zhuzilin committed:
remove redundant comment and backslash

zhuzilin commented on tensorflow/python/keras/optimizers_test.py:
```diff
@@ -232,6 +233,24 @@ def test_negative_clipvalue_or_clipnorm(self):
+    model = keras.models.Sequential()
+    model.add(keras.layers.Dense(
+        2, input_shape=(3,), kernel_constraint=keras.constraints.MaxNorm(1)))
+    # This is possible
```
Removed.

reedwm commented:
Thanks for the fix!

zhuzilin committed:
add indentation

zhuzilin commented:
@reedwm Just fixed the indentation error around the changed parenthesis. Could you please approve it again? Thank you for your time!

zhuzilin commented:
@reedwm Could you help me merge this pr? Thank you!
